efi/env.go

```diff
@@ -0,0 +1,305 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019-2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package efi
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha1"
+	"encoding/binary"
+	"io"
+	"sort"
+
+	efi "github.com/canonical/go-efilib"
+	"github.com/canonical/tcglog-parser"
+	"golang.org/x/xerrors"
+)
+
+// HostEnvironment is an interface that abstracts out an EFI environment, so that
+// consumers of the API can provide a custom mechanism to read EFI variables or parse
+// the TCG event log.
+type HostEnvironment interface {
+	// ReadVar reads the specified EFI variable
+	ReadVar(name string, guid efi.GUID) ([]byte, efi.VariableAttributes, error)
+
+	// ReadEventLog reads the TCG event log
+	ReadEventLog() (*tcglog.Log, error)
+}
+
+// varReader is a subset of HostEnvironment that is just for EFI variables
+type varReader interface {
+	// ReadVar reads the specified EFI variable
+	ReadVar(name string, guid efi.GUID) ([]byte, efi.VariableAttributes, error)
+}
+
+// rootVarReaderKey is a SHA1 used to uniquely identify the contents of a rootVarReader.
+// This is to ensure that we don't generate multiple profile branches with the same
+// starting state.
+type rootVarReaderKey [sha1.Size]byte
+
+type varContents struct {
+	attrs efi.VariableAttributes
+	data  []byte
+}
+
+// rootVarReader provides the initial state for each profile generation branch. It
+// consists of the host environment and updates that have been generated by previous
+// iterations of the profile generation.
```

Comment:
The comment should be able to point to some code in a future PR. Essentially the public function `AddPCRProfile` will add a branch point to the supplied `PCRProtectionProfileBranch` and it will then run a loop which adds a full profile branch for each collected environment (where the first environment is based on the initial `HostEnvironment` using the supplied `[]ImageLoadActivity`.